modules/standard/IO.chpl

```diff
@@ -5229,6 +5229,149 @@ proc _channel.writeBinary(arg:numeric, endian:ioendian) throws {
+
+   :arg s: the ``string`` to write
+   :arg size: the number of codepoints to write from the ``string``
+
+   :throws SystemError: Thrown if the string could not be written to the channel.
+   :throws IllegalArgumentError: Thrown if ``size`` is larger than ``s.size``
+*/
+proc _channel.writeBinary(s: string, size: int = s.size) throws {
+  // handle bad arguments
+  if size > s.size then
+    throw new owned IllegalArgumentError("size", "cannot exceed length of provided string");
+  if s.hasEscapes then
+    throw createSystemError(EILSEQ, "illegal use of escaped string characters in 'writeBinary'");
+
+  // count the number of bytes to write
+  var sLocal = s.localize();
+  var bytesLen = 0;
+  if size == sLocal.size {
+    bytesLen = s.numBytes;
+  } else {
+    for ((_, cp_byte_len), i) in zip(sLocal._indexLen(), 0..) {
+      if i == size then break;
+      bytesLen += cp_byte_len;
+    }
+  }
+
+  // write the first bytesLen bytes of the string to the channel
+  var e: errorCode = qio_channel_write_string(
+    false,
+    iokind.native: c_int,
+    qio_channel_str_style(this._channel_internal),
+    this._channel_internal,
+    sLocal.c_str(),
+    bytesLen: c_ssize_t
+  );
+
+  if e != 0 then
+    throw createSystemError(e);
+}
+
+/*
+   Write a :type:`~Bytes.bytes` to the channel in binary format
+
+   :arg s: the ``bytes`` to write
+   :arg size: the number of bytes to write from the ``bytes``
+
+   :throws SystemError: Thrown if the bytes could not be written to the channel.
+   :throws IllegalArgumentError: Thrown if ``size`` is larger than ``b.size``
+*/
+proc _channel.writeBinary(b: bytes, size: int = b.size) throws {
+  // handle bad arguments
+  if size > b.size then
+    throw new owned IllegalArgumentError("size", "cannot exceed length of provided bytes");
+
+  // write the first size bytes to the channel
+  var bLocal = b.localize();
+  var e: errorCode = qio_channel_write_string(
+    false,
+    iokind.native: c_int,
+    qio_channel_str_style(this._channel_internal),
+    this._channel_internal,
+    bLocal.c_str(),
+    size: c_ssize_t
+  );
+
+  if e != 0 then
+    throw createSystemError(e);
+}
+
+/*
+   Write an array of binary numbers to the channel
+
+   Note that this routine currently requires a 1D rectangular non-strided array.
+
+   :arg data: an array of numbers to write to the cannel
+   :arg endian: :type:`ioendian` compile-time argument that specifies the byte order in which
+              to read the numbers. Defaults to ``ioendian.native``.
+
+   :throws SystemError: Thrown if the values could not be written to the channel.
+   :throws UnexpectedEofError: Thrown if EOF is reached before all the values could be written.
+*/
+proc _channel.writeBinary(const ref data: [?d] ?t, param endian:ioendian = ioendian.native) throws
+  where (d.rank == 1 && d.stridable == false) && (
+          isIntegralType(t) || isRealType(t) || isImagType(t) || isComplexType(t))
+{
+  var e : errorCode = 0;
+
+  on this._home {
```

Comment:
I'm noticing that this version uses an `on` statement while the previous ones localize the string to whatever locale the program is currently on.  Is that intentional?

Reply:
I went with `string/bytes.localize()` because that's what `_write_binary_internal` uses for strings and bytes. Would it be better to switch those to use `on` statements instead?

Reply:
I don't actually know.  I suspect probably following `_write_binary_internal`'s lead is the right call, but I'd be curious to hear @mppf's take when he gets the chance